src/python/dxpy/bindings/dxfile_functions.py

```diff
@@ -466,3 +466,101 @@ def upload_string(to_upload, media_type=None, keep_open=False, wait_on_close=Fal
         handler.close(block=wait_on_close, **remaining_kwargs)
 
     return handler
+
+def static_var(var_name, initial_value):
+    def decorate(func):
+        setattr(func, var_name, initial_value)
+        return func
+    return decorate
+
+@static_var("folders_cache", {})
+def list_subfolders(project, path, usecache=True, recurse=True):
+    '''
+    :param project: Project ID to use as context for the listing
+    :type project: string
+    :param path: Subtree root path
+    :type path: string
+    :param usecache: Use folders listing cache
+    :type usecache: boolean
+    :param recurse: Return a complete subfolders tree
+    :type recurse: boolean
+
+    Returns a list of subfolders for the remote *path* (included to the result) of the *project*.
+
+    Example::
+
+        list_subfolders("project-xxxx", folder="/input", usecache=False)
+
+    '''
+    if usecache:
+        if project not in list_subfolders.folders_cache:
+            list_subfolders.folders_cache[project] = dxpy.get_handler(project).describe(input_params={'folders': True})['folders']
+        project_folders = list_subfolders.folders_cache[project]
+    else:
+        project_folders = dxpy.get_handler(project).describe(input_params={'folders': True})['folders']
+    # TODO: support shell-style path globbing (i.e. /a*/c matches /ab/c but not /a/b/c)
+    # return pathmatch.filter(project_folders, os.path.join(path, '*'))
+    if recurse:
+        return (f for f in project_folders if f.startswith(path))
+    else:
+        return (f for f in project_folders if f.startswith(path) and '/' not in f[len(path)+1:])
+
+def download_folder(project, destdir, folder="/", overwrite=False, chunksize=dxfile.DEFAULT_BUFFER_SIZE,
+        usecache=True, **kwargs):
+    '''
+    :param project: Project ID to use as context for this download.
+    :type project: string
+    :param destdir: Local destination location
+    :type destdir: string
+    :param folder: Path to the remote folder to download
+    :type folder: string
+    :param overwrite: Overwrite existing files
+    :type overwrite: boolean
+    :param usecache: Use folders listing cache
+    :type usecache: boolean
+
+    Downloads a remote *folder* of the *project* and saves it to the *destdir* local location.
+
+    Example::
+
+        download_folder("project-xxxx", "/home/jsmith/input", folder="/input")
+
+    '''
+
+    def ensure_local_dir(d):
+        if not os.path.isdir(d):
+            if os.path.exists(d):
+                raise DXFileError("Destination location '{}' already exists and is not a directory".format(destdir))
+            logger.debug("Creating destination directory: '%s'", d)
+            os.makedirs(d)
+
+    def compose_local_dir(remote_subfolder):
+        return os.path.normpath(os.path.join(destdir, remote_subfolder[1:] if folder == "/" else remote_subfolder[len(folder) + 1:]))
+
+    folder = os.path.normpath(folder).strip()
```

Comment:
I'm not sure if it is permissible to use "os.path" here, since the paths will be handled using the system conventions. That is, this code may be running on Windows but `folder` must be interpreted like a unixy path regardless. Would it be sufficient to strip a trailing "/" here, or do you need more than that?